File: 647-palindromic-substrings/palindromic-substrings.cpp
#include <string>
using namespace std;

class Solution {
public:
    int count = 0;

    bool isPalindrome(const string& str, int start, int end) {
        while (start < end) {
            if (str[start++] != str[end--]) {
                return false;
            }
        }
        return true;
    }

    void backtrack(const string& s, int start) {
        if (start == s.length()) {
            return;
        }

        for (int end = start; end < s.length(); end++) {
            if (isPalindrome(s, start, end)) {
                count++;
            }
        }

        backtrack(s, start + 1); // Move to the next starting index
    }

    int countSubstrings(string s) {
        backtrack(s, 0);
        return count;
    }
};

File: 647-palindromic-substrings/palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        count = 0 
        n = len(s)
        for i in range(n):
            l, r = i,i
            while l>=0 and r < n and l <=r and s[l] == s[r]:
                count += 1
                l -= 1
                r += 1
            l, r = i, i + 1
            while l >= 0 and r < n and l <= r and s[l] == s[r]:
                count += 1
                l -=1 
                r += 1
        return count 